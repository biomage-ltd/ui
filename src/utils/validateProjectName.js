const defaultErrorMessages = {
  MIN_8: 'Min 8 characters',
  MIN_2_CHAR: 'Min 2 characters in sequence',
  INVALID_CHARS: 'Invalid characters',
  NAME_EXISTS: 'A project with the same name exists',
};

const validateProjectName = (input, projectNames, customErrors) => {
  const errorMessages = {
    ...defaultErrorMessages,
    customErrors,
  };

  if (input.length < 8) {
    return errorMessages.MIN_8;
  }

  if (!input.match(/([a-zA-Z\d]{2,}){1,}/gm)) {
    return errorMessages.MIN_2_CHAR;
  }

  if (!input.match(/^[a-zA-Z\s\d-_]{8,}$/gm)) {
    return errorMessages.INVALID_CHARS;
  }

  if (projectNames instanceof Array
    && projectNames.includes(input)) {
    return errorMessages.NAME_EXISTS;
  }

  if (projectNames instanceof Set
    && projectNames.has(input)) {
    return errorMessages.NAME_EXISTS;
  }

  return true;
};

export default validateProjectName;
