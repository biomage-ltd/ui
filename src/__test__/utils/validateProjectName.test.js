import validateProjectName from '../../utils/validateProjectName';

describe('validateProjectName unit test', () => {
  it('project names minimal 8 char ', () => {
    const validName = 'Project 1';

    expect(validateProjectName(validName)).toEqual(true);

    const inValidName = 'asd';
    expect(validateProjectName(inValidName)).not.toEqual(true);
    expect(validateProjectName(inValidName)).toMatchSnapshot();
  });

  it('project names has to at least have two consequent letters', () => {
    const validName = ' ab  cd  ef  gh';
    expect(validateProjectName(validName)).toEqual(true);

    const inValidName = 'a b c d e f g h';
    expect(validateProjectName(inValidName)).not.toEqual(false);
    expect(validateProjectName(inValidName)).toMatchSnapshot();
  });

  it('project name should not contain invalid characters', () => {
    const invalidName = 'Yumm: A great project!';

    expect(validateProjectName(invalidName)).not.toEqual(true);
    expect(validateProjectName(invalidName)).toMatchSnapshot();
  });

  it('project name should not be the same with existing projects', () => {
    const validName = 'Project 4';
    const projectNames = [
      'Project 1',
      'Project 2',
      'Project 3',
    ];
    const inValidName = projectNames[0];

    expect(validateProjectName(validName, projectNames)).toEqual(true);
    expect(validateProjectName(inValidName, projectNames)).not.toEqual(false);
    expect(validateProjectName(inValidName)).toMatchSnapshot();
  });
});
