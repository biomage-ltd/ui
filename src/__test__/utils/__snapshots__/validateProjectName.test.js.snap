// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateProjectName unit test project name should not be the same with existing projects 1`] = `true`;

exports[`validateProjectName unit test project name should not contain invalid characters 1`] = `"Only letters, numbers, space, _, and - allowed"`;

exports[`validateProjectName unit test project names has to at least have two consequent letters 1`] = `"Min. 2 characters in sequence"`;

exports[`validateProjectName unit test project names minimal 8 char  1`] = `"Min. 8 characters required"`;
